### PR TITLE
[SPARK-40538] [CONNECT] Improve built-in function support for Python client.

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
This patch changes the way simple scalar built-in functions are resolved in the Python Spark Connect client. Previously, it was trying to manually load specific functions. With the changes in this patch, the trivial binary operators like `<`, `+`, ... are mapped to their name equivalents in Spark so that the dynamic function lookup works.

In addition, it cleans up the Scala planner side to remove the now unnecessary code translating the trivial binary expressions into their equivalent functions.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT, E2E

### CODE REVIEW

**Comment by grundprinzip** [comment]
R: @hvanhovell @cloud-fan @HyukjinKwon @amaliujia

**Review by hvanhovell — APPROVED** [review]
LGTM

File: python/pyspark/sql/connect/column.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def to_plan(self, session: Optional["RemoteSparkSession"]) -> proto.Expression:
         return self.ref.to_plan(session)
 
 
-class ScalarFunctionExpression(Expression):
+class ScalarFunctionExpression(ColumnRef):
```
Comment (amaliujia) [inline python/pyspark/sql/connect/column.py:164]: Why a function expression inherit from ColumnRef now?

**Comment by amaliujia** [comment]
Can you also improve the Scala side testing coverage as you have update the SparkConnectPlanner?

**Comment by grundprinzip** [comment]
> Can you also improve the Scala side testing coverage as you have updated the SparkConnectPlanner?

I'll figure out to add more. Interestingly the work here is already covered by your previous patch on the where clause. Because we rely on the unresolved function call. But I'll add a negative test case.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala [hunk 2]
```diff
@@ -193,20 +189,10 @@ class SparkConnectPlanner(plan: proto.Relation, session: SparkSession) {
...
-      case "eq" =>
-        assert(fun.getArgumentsCount == 2, "`eq` function must have two arguments.")
-        expressions.EqualTo(
-          transformExpression(fun.getArguments(0)),
-          transformExpression(fun.getArguments(1)))
-      case _ =>
-        lookupFunction(funName, fun.getArgumentsList.asScala.map(transformExpression).toSeq)
-    }
+    UnresolvedFunction(
+      Seq(funName),
```
Comment (amaliujia) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:193]: Testing a negative case is great. 

I was also thinking a test here:
```
val funName = fun.getPartsList.asScala.mkString(".")

UnresolvedFunction(
      Seq(funName),
      ...
```

Basically it first concatenates the function name by `.` then passing that to a Seq which become Seq with one element, but why not just don't do the concatenation, or we have to do the  concatenation?

Having a test case here will better demonstrate the intention of this implementation.

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:193]: Will do!

File: python/pyspark/sql/connect/column.py (hunk 1 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:164]: I'll refactor this to make the difference between Expression and ColumnRef clear. Essentially, I'll move all of the binary operators up to the Expression class.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala (hunk 2 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala:193]: There is another thing that we probably have to do similar to what you've done for the unresolved identifier, just passing the full string to the backend and then figure out there when do we split accordingly.

**Comment by grundprinzip** [comment]
@amaliujia it would be great if you could have another look. I added additional tests on the planner side to iron out the issue with the name parts. 

Thanks

File: python/pyspark/sql/tests/connect/test_connect_basic.py [hunk 3]
```diff
@@ -19,9 +19,12 @@
 import unittest
 import tempfile
 
+import pandas
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_connect_basic.py:22]: Hm .. we gonna fix this or do something. pandas isn't a required library for SQL package. Should probably skip this tests when pandas is not installed for now until we have a clear way to handle this. (see `pyspark.testing.sqlutils.have_pandas` and `pyspark.sql.tests.test_arrow_map`

File: python/pyspark/sql/connect/column.py [hunk 4]
```diff
@@ -19,18 +19,59 @@
 
 
 import pyspark.sql.connect.proto as proto
+
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/column.py:22]: Should probably remove this since `pyspark` should be within the same import block.

File: python/pyspark/sql/connect/column.py [hunk 5]
```diff
@@ -19,18 +19,59 @@
...
+    return _
+
+
 class Expression(object):
     """
     Expression base class.
     """
 
+    __gt__ = _bin_op(">")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/column.py:49]: Why was this moved from Column to Expression?

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala [hunk 6]
```diff
@@ -64,6 +64,36 @@ package object dsl {
...
     }
 
+    /**
+     * Create an unresolved function from name parts.
+     *
+     * @param nameParts
+     * @param args
+     * @return Expression wrapping the unresolved function.
+     */
+    def fun(nameParts: Seq[String], args: Seq[proto.Expression]): proto.Expression = {
```
Comment (HyukjinKwon) [inline connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:74]: I would name it `func` as that's (much) more common in the codebase as far as I can tell.

File: python/pyspark/sql/connect/column.py (hunk 5 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:49]: Without this change the different expressions are not combinable. Until now only column references were combinable but actually all column expressions should be.

File: python/pyspark/sql/connect/column.py [hunk 7]
```diff
@@ -73,20 +114,6 @@ def __str__(self) -> str:
...
-            other = LiteralExpression(other)
-        if not reverse:
-            return ScalarFunctionExpression(name, self, other)
-        else:
-            return ScalarFunctionExpression(name, other, self)
-
-    return _
-
-
 class ColumnRef(Expression):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/column.py:117]: I think we should rename this to `AttributeReference` to avoid confusion e.g., I think it's a bit mixed with `Column` interface that is the user-facing interface. @amaliujia and @cloud-fan 

Should better to keep it matched with either Catalyst internal types or user-facing Spark SQL interface classes.

File: python/pyspark/sql/tests/connect/test_connect_basic.py (hunk 3 above)
Comment (grundprinzip) [reply python/pyspark/sql/tests/connect/test_connect_basic.py:22]: Interestingly, nothing in Spark Connect will work atm without pandas because we always call `toPandas` in the collection of the result. Let me know what you want to do.

File: python/pyspark/sql/connect/column.py (hunk 4 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:22]: yes, will clean it up

File: python/pyspark/sql/connect/column.py (hunk 7 above)
Comment (grundprinzip) [reply python/pyspark/sql/connect/column.py:117]: Let's have a discussion about this, but this is an unrelated change to this one. I think we should probably call `Expression` -> `Column` and `ColumnRef` -> `AttributeReference` but it will require some more digging what the right name should be. However, as said, that's independent of this change.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala (hunk 6 above)
Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:74]: this is just in the DSL package that is used for testing. I'll check what Catalyst is doing.

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:74]: the catalyst DSL has a similar `callFunction`

Comment (HyukjinKwon) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:74]: For a bit of more context, one thing is that we use `snake_case` to match w/ SQL function names (see `Column` or `functions.scala`). This kind of naming rule is already mixed in our existing SQL DSL (see also `org.apache.spark.sql.catalyst.package`). Should probably pick one and stick to that.

In the past, we followed `camelCase` in both DSL, `Column` and `functions.scala`. After that, we renamed them all to `snake_case` for SQL compatibility in `Column` and `functions.scala` (so the new DSL added follows  `snake_case` at `org.apache.spark.sql.catalyst.package`)

Therefore, I tend to use `snake_case` in this DSL case too but I am fine if others (or you) feel strongly this is better.

**Review by HyukjinKwon — APPROVED** [review]
One comment: https://github.com/apache/spark/pull/38270#discussion_r996537113. LGTM since we're still heavily developing but should probably revisit things like https://github.com/apache/spark/pull/38270#discussion_r996476173 or https://github.com/apache/spark/pull/38270#discussion_r996476445

File: python/pyspark/sql/connect/column.py (hunk 7 above)
Comment (amaliujia) [reply python/pyspark/sql/connect/column.py:116]: +1 I have been thinking this `ColumnRef` thing. Let's revisit it on the naming, etc.

File: connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala (hunk 6 above)
Comment (zhengruifeng) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:74]: maybe name the two methods `function`?
 https://github.com/apache/spark/blob/ff7ab34a6957965d74504ed1d36de21d1ad7319c/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala#L368-L369

don't feel strong about the naming

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:74]: Oh nice, I definitely missed renaming the other overall `callFunction` as well. TBH I'm not sure what the right approach is here because the catalyst DSL calls it `callFunction` instead of `call_function` 🤷

Comment (grundprinzip) [reply connector/connect/src/main/scala/org/apache/spark/sql/connect/dsl/package.scala:74]: https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/dsl/package.scala#L250-L256 Please let me know what you want to do.

**Comment by HyukjinKwon** [comment]
Merged to master.

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala [hunk 8]
```diff
@@ -50,6 +50,34 @@ class SparkConnectProtoSuite extends PlanTest with SparkConnectPlanTest {
...
+    assertThrows[UnsupportedOperationException] {
+      connectPlan.analyze
+    }
+
+    val validPlan = {
+      import org.apache.spark.sql.connect.dsl.expressions._
+      import org.apache.spark.sql.connect.dsl.plans._
+      transform(connectTestRelation.select(callFunction(Seq("hex"), Seq("id".protoAttr))))
+    }
+    assert(validPlan.analyze != null)
```
Comment (cloud-fan) [inline connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:78]: it's better to compare it with the catalyst plan

**Comment by cloud-fan** [comment]
late LGTM

File: connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala (hunk 8 above)
Comment (grundprinzip) [reply connector/connect/src/test/scala/org/apache/spark/sql/connect/planner/SparkConnectProtoSuite.scala:78]: This is not to validate that the catalyst plan exists, but really just that existing functions are actually resolved. The `!=null` is mostly to have any assertion and not throw.

File: python/pyspark/sql/connect/column.py [hunk 9]
```diff
@@ -26,11 +26,51 @@
...
+    return _
+
+
 class Expression(object):
     """
     Expression base class.
     """
 
+    __gt__ = _bin_op(">")
+    __lt__ = _bin_op(">")
```
Comment (cloud-fan) [inline python/pyspark/sql/connect/column.py:49]: `_bin_op("<")`

Comment (HyukjinKwon) [reply python/pyspark/sql/connect/column.py:49]: Yeah, I think this was a mistake.

Comment (HyukjinKwon) [reply python/pyspark/sql/connect/column.py:49]: https://github.com/apache/spark/pull/38303
